USI Mock implementation

timcharper opened the pull request:
In this PR we implement a basic mocked implementation of the USI. One can provide the mocked mesos flow as seen in the unit test and use this "mock interface" to "launch pods".

JIRA Issues: DCOS-47473

Commit:
Mock USI implementation

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/Goal.scala:
```diff
@@ -0,0 +1,8 @@
+
+sealed trait Goal
+
+object Goal {
```
I guess I'm a little repetitive but please add docs 🙂

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/PodRecord.scala:
```diff
@@ -1,8 +1,11 @@
   */
-case class PodRecord(podId: PodId)
\ No newline at end of file
+case class PodRecord(podId: PodId, launchedAt: Instant, agentId: AgentId)
```
What does `launchedAt` mean? Could you add the `@param` docs?

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/PodSpec.scala:
```diff
@@ -4,4 +4,4 @@ package com.mesosphere.usi.core.models
   */
-case class PodSpec(id: PodId)
\ No newline at end of file
+case class PodSpec(id: PodId, goal: Goal, runSpec: String)
```
`@param` docs are missing. Should we type `runSpec`?

jeschkies commented on core/build.gradle:
```diff
@@ -1,3 +1,5 @@
 dependencies {
     compile project(':core-models')
+
+    compile 'com.typesafe.akka:akka-stream_2.12:2.5.19'
```
We have a global `scalaVersion` defined.

jeschkies replied:
```suggestion
    compile "com.typesafe.akka:akka-stream_$scalaVersion:2.5.19"
```

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/Response.scala:
```diff
@@ -0,0 +1,34 @@
+
+import com.mesosphere.usi.models.{Mesos, StatusMessage}
+
+sealed trait Response {
```
What is a response?

alenkacz replied:
yeah I have problems with naming in this part as well... :/ I think we should find better name for this class

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/Response.scala:
```diff
@@ -0,0 +1,34 @@
+
+case class ResponseWithMessages(statusMessages: List[StatusMessage] = Nil, mesosCalls: List[Mesos.Call] = Nil) extends Response
+
+case class ResponseBuilder(reverseStatusMessage: List[StatusMessage] = Nil, reverseMesosCalls: List[Mesos.Call] = Nil) extends Response {
```
Maybe I misunderstand something but should the `ResponseBuilder` have a `build` method?

timcharper replied:
It does not need one

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -1,14 +1,27 @@
 
-import scala.concurrent.Future
+object Scheduler {
+  def graph: BidiFlow[SpecMessage, StatusMessage, Mesos.Event, Mesos.Call, NotUsed] = {
```
Could you add docs to this method? It seems to be of importance but I have no idea what's really happening here?

alenkacz replied:
it helped me to draw it down

```    * {{{
    *                 +------+
    *  SpecMessages ~>|      |~> StatusMessages
    *                 | bidi |
    *  Mesos.Call <~  |      |<~ Mesos.Event
    *                 +------+
    * }}}
```

but I still don't totally get the flow of data here. I would expect SpecMessage triggers Mesos.Call and Mesos.Event triggers StatusMessage but this does not seem to be the case right now. @timcharper

alenkacz replied:
also I guess this is not the most highest level interface, right? The one framework will interact... Because then I would expect it to be just `SpecMessages ~> StatusMessages` I think I am still puzzled by the fact that we expose things that are to me on totally different levels in one component - that's the outside USI interface as well as stuff happening internally (mesos calls and events). Could you maybe clarify that?

jeschkies replied:
+1

The visualization helps a lot.

alenkacz replied:
I still believe the wrapper handling the lifecycle of mesos client should be part of the USI. We don't have to do it in this PR but I kind of strongly feel that this should be handled by the library.

I really like the separation of concerns and I think we're not gonna lose it if we include the wrapper simplifying this to `SpecMessages ~> StatusMessages`. That to me is the interface I would want to use from a framework perspective when using scheduling library

alenkacz replied:
also please put the ascii art from slack in here as a scaladoc

jeschkies replied:
Could we add the flow chart to the docs above. If it uses PlantUML or Mermaid syntax I'm more than happy to render them in our scaladocs.

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+object SchedulerLogic {
+  private def assertValidTransition(oldSpec: PodSpec, newSpec: PodSpec): Unit = {
+    if ((oldSpec.goal == Goal.Terminal) && (newSpec.goal == Goal.Running))
+      throw new IllegalStateException(s"Illegal state transition Terminal to Running for podId ${oldSpec.id}")
```
How are exceptions handled and where? Should this bubble up to the orchestration layer?

timcharper replied:
Ideally we'd crash.

Also, this is a mock implementation.

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+  }
+
+
+  def matchOffer(podSpecs: Map[PodId, PodSpec])(offer: Mesos.Event.Offer, pendingLaunch: Set[PodId], responseBuilder: ResponseBuilder): ResponseBuilder = {
```
I like that this is a pure function. What is the motivation for using the builder pattern?

timcharper replied:
We discussed this in a meeting

jeschkies commented on core/src/test/scala/com/mesosphere/usi/core/AkkaUnitTest.scala:
```diff
@@ -0,0 +1,14 @@
+import org.scalatest.concurrent.ScalaFutures
+import org.scalatest.{BeforeAndAfterAll, Suite}
+
+trait AkkaUnitTest extends Suite with BeforeAndAfterAll with ScalaFutures {
```
See `test-utils` from @zen-dog 🙂

alenkacz replied:
how is that related? since this is unit test? 🤔

jeschkies replied:
Alex imported `AkkaUnitTest` as well. `test-utils` is not just for integration tests.

jeschkies reviewed (commented):
I know this is an early draft so bear with my comments 🙂 

Overall, I had difficulties finding the actual interface. I don't know the entry point into the scheduler.

I was also wondering if we really neat complex graphs. I'm probably missing a lot but I expected something like the Elm/Redux structure where we would fold over a stream of events and the events being Mesos updates or PodSpec changes.

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -1,14 +1,27 @@
+          val mesosCalls = broadcast.out(0).mapConcat { response => response.mesosCalls }
+          val statuses = broadcast.out(1).mapConcat { response => response.statusMessages }
+
+          BidiShape.apply(schedulerLogic.in0, statuses.outlet, schedulerLogic.in1, mesosCalls.outlet)
```
can we create variables for in0 and in1 to make it easier to read? something like `val specMessages = schedulerLogic.in0` and then using that in `BidiShape.apply`

kaiwalyajoshi replied:
Seconded, even better if `schedulerLogic.in0` and `schedulerLogic.in1` can match `MesosEvents` and `FrameworkIn` from the diagram.

timcharper replied:
Let's tackle that later; I think this is pretty okay to read and the type system has our back here.

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala:
```diff
@@ -0,0 +1,81 @@
+  val specMessages = Inlet[SpecMessage]("specs")
+  val responses = Outlet[Response]("response")
+  // Define the (sole) output port of this stage
+  val out: Outlet[Int] = Outlet("NumbersSource")
```
I think you did not mean to have the `NumbersSource` here :)

timcharper replied:
✅

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala:
```diff
@@ -0,0 +1,81 @@
+
+  val mesosEvents = Inlet[Mesos.Event]("mesos-events")
+  val specMessages = Inlet[SpecMessage]("specs")
+  val responses = Outlet[Response]("response")
```
can we find better name for Response? To me even something as simple as `SchedulerLogicOutput` would be better...

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala:
```diff
@@ -0,0 +1,81 @@
+      def handleResponse(response: Response): Unit = {
+        if (isAvailable(responses)) {
+          if (pendingResponses.nonEmpty) {
+            throw new IllegalStateException("We should always immediately push on pull if responses are queued")
```
this part I don't understand... why is this state illegal? Especially since on l.66 you do `pendingResponses.enqueue(response)` - how do we ensure that next time we don't end up at this place with pendingResponses containing a value? 🤔

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala:
```diff
@@ -0,0 +1,81 @@
+object SchedulerLogicGraph {
+  val BUFFER_SIZE = 32
+}
+class SchedulerLogicGraph extends GraphStage[FanInShape2[SpecMessage, Mesos.Event, Response]] {
```
I think this graph deserves ASCII art as well

alenkacz replied:
so basically this piece of logic is just scheduler processing with added backpressure, correct? Could we put that into docs and maybe even rename the class? One has to figure out now by reading the code :)

alenkacz commented on core/src/main/scala/com/mesosphere/usi/models/Mesos.scala:
```diff
@@ -0,0 +1,53 @@
+/**
+  * Mock Mesos calls / state
+  */
+object Mesos {
```
I guess this one will go when we put in the mesos client implementation

zen-dog replied:
It's in now ;)

timcharper replied:
Let's integrate the mesos client in the task specifically designated as such

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala:
```diff
@@ -0,0 +1,81 @@
+import scala.collection.mutable
+
+object SchedulerLogicGraph {
+  val BUFFER_SIZE = 32
```
should this in the end be exposed as a config value for orchestrator to increase/decrease based on their environment?

zen-dog replied:
Absolutely. We already have an established pattern of using typesafe config (with Settings* classes) - this should be used here too.

timcharper replied:
Lets do it later. Mock interface.

alenkacz commented on core/src/test/scala/com/mesosphere/usi/core/SchedulerTest.scala:
```diff
@@ -0,0 +1,65 @@
+
+    input.offer(PodSpecChange(podId, Some(PodSpec(podId, Goal.Running, RunSpec(1)))))
+    inside(output.pull().futureValue) {
+      case Some(podRecord: PodRecordChange) =>
```
why not `PodSpecChange`? It's a little bit weird that I put in PodSpec and get PodRecord... I always thought that I should get back PodSpecChange... and records are just implementation detail of the scheduler

alenkacz commented on core/src/test/scala/com/mesosphere/usi/core/SchedulerTest.scala:
```diff
@@ -0,0 +1,65 @@
+      .toMat(Sink.queue())(Keep.both)
+      .run
+
+    input.offer(PodSpecChange(podId, Some(PodSpec(podId, Goal.Running, RunSpec(1)))))
```
shouldn't snapshot be always first? 🤔

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+  }
+
+  def handleSpecMessage(msg: SpecMessage): Response = {
+    val handleResponse = SchedulerLogic.handleSpecMessage(this.podRecords)(this.podSpecs, msg)
```
why isn't this `(podRecords, podSpecs)(msg)`?

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+  private def podIdFor(taskId: TaskId): PodId = PodId(taskId)
+}
+
+class SchedulerLogic {
```
I am still struggling - when do you put a method here and when in `SchedulerLogic` object? In class are only functions that modify the state?

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+    * We remove a task if it is not reachable and running, and it has no podSpec defined
+    *
+    * Should be called with the effects already applied for the specified podIds
+    * @param podIds podIds changed during the last frame
```
you're refering to a frame here but we don't have frames anywhere. I think we should be consistent with naming...

timcharper replied:
Frames, as a concept, still exists

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+    b
+  }
+
+  def pendingLaunch(podId: PodId, goal: Goal, podRecord: Option[PodRecord]): Boolean = {
```
podId is not used. Also `isPendingLaunch`?

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+  private def updatePendingLaunch(podIds: Set[PodId]): Unit = {
+    podIds.foreach { podId =>
+      val shouldBeLaunched = podSpecs.get(podId).exists { podSpec =>
+        SchedulerLogic.pendingLaunch(podId, podSpec.goal, podRecords.get(podId))
```
can we do `import SchedulerLogic._` please? It would make stuff easier to read...

timcharper replied:
There's not that much here, I generally dislike splat imports. See if you still feel that way? I've moved a lot of logic out.

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+
+    newState match {
+      case None =>
+        b = b.withChangeMessage(PodStatusChange(podId, None))
```
I don't understand, why do we need two here? Why is `PodStatusChange` not enough?

timcharper replied:
Why do we need two what here? PodStatus and PodRecord?

Because the records are updated independently. A PodStatus may not have a PodRecord, and vice versa. When we launch a pod, it has no status yet, but we must have a record that we launched it. And, if we discover a status for which there is no record, then its a spurious task.

If we coupled task status with pod record then we would needlessly need to rewrite the podRecord every time a Mesos health check was received or whatnot.

alenkacz commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+        podSpecs.get(podId).foreach(assertValidTransition(_, podSpec))
+
+        if (podSpec.goal == Goal.Terminal) {
+          taskIdsFor(podSpec).foreach { taskId =>
```
I hope we're gonna extract the killing logic out of here later. E.g. I don't think we should kill every time you receive update. Do we have JIRA to implement proper killing?

timcharper replied:
Yes, that is the intention. I can add a comment make that more clear. No specific JIRA, but, then again, were not even doing proper test killing anyways. As an aside, is there really a downside to sending a kill on every terminal update? It's kind of an idempotent thing.

alenkacz reviewed (approved):
Uff, I spent several hours going through this and I still don't feel like I am done. What would help me a lot at this stage is if some things were not as deep as they are (e.g. we have killing to mesos implemented as well as revive or pruning tasks. That's just too low level for me at this point when I want to focus on the shape of the overall interface). If this was only graphs + very dummy logic, it would be much easier.

Overall I think we have the biggest problem with naming, we have a lot of methods like "handleXYZ" calling one another and just looking at the method it's not always apparent what is it supposed to do.  E.g. `handlePodSpecUpdate` and `handleSpecMessage` where the second is handling `PodSpecChange`. What's the difference between update and change.

Also I think we definitely need better name for `Response`.

I am on vacation next week so I don't want to block this and at the same time I think it's not set in stone and we can still improve on that so marked it as ✅

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/Goal.scala:
```diff
@@ -0,0 +1,19 @@
+package com.mesosphere.usi.core.models
+
+/**
+  * Target state of PodSpec. Scheduler will work towards achieving this goal state.
```
```suggestion
  * Target state of [[PodSpec]]. Scheduler will work towards achieving this goal state.
```

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/Goal.scala:
```diff
@@ -0,0 +1,19 @@
+object Goal {
+
+  /**
+    * PodSpec with this goal should be started if is not running
```
```suggestion
    * [[PodSpec]] with this goal should be started if is not running
```

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/Goal.scala:
```diff
@@ -0,0 +1,19 @@
+  case object Running extends Goal
+
+  /**
+    * PodSpec with this goal should be killed if it is running.
```
```suggestion
    * [[PodSpec]] with this goal should be killed if it is running.
```

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/PodRecord.scala:
```diff
@@ -1,8 +1,13 @@
+  *
   * @param podId id of the pod
+  * @param launchedAt time at which we accepted an offer and initiated launch of this pod
+  * @param agentId id of mesos agent from which we accepted offer to launch this pod
```
```suggestion
  * @param agentId id of Mesos agent from which we accepted offer to launch this pod
```

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,251 @@
+
+  var cachedPendingLaunch: Set[PodId] = Set.empty
+
+  private def applyChanges(response: Response): Unit = {
```
This should have some docs. I understand the `response` basically records all changes we want to have. Is that correct?

jeschkies reviewed (commented):
I slowly get an idea of the flow. I think all should have a go at this code before we merge.

I'm a little unsure about the response builder. I like it on principles but we might run into issues soon.

In my [refactoring](https://github.com/mesosphere/marathon/pull/6746) I've found that it was not enough to just record changes since some logic further down would require an already update state. Eg in [TaskReplaceActorBehaviour.process](https://github.com/mesosphere/marathon/pull/6746/files#diff-799d40bf745dd8a6faf58e52993998b8R72) we have
```scala
def process(completedPhases: Int, frame: Frame): FrameProcessor.ProcessResult = {
    if (check(frame)) {
      FrameProcessor.Stop
    } else {
      val frameAfterKilling = if (completedPhases == 0) {
        killImmediately(ignitionStrategy.nrToKillImmediately, frame)
      } else {
        killNext(frame)
      }
      val nextFrame = launching(frameAfterKilling)
      FrameProcessor.Continue(nextFrame)
    }
  }
  }
```

`killNext` and `launching` return a new `Frame`, ie a new state. `launching` needs to see the changes from `killNext`. If `killNext` would "just" return the queued responses `launching` would have to apply them first. That was tedious so I changed the [Frame to do that for me](https://github.com/mesosphere/marathon/pull/6746/files#diff-41d67b6cd090ef8b479199b2786fde10R54). I'm not saying my solution is ideal but it made life easier for the core logic. In essence the `Frame` is a builder that also has the intermediate state.

zen-dog commented on core/build.gradle:
```diff
@@ -1,3 +1,5 @@
 dependencies {
     compile project(':core-models')
+
+    compile 'com.typesafe.akka:akka-stream_2.12:2.5.19'
```
```suggestion
```

zen-dog replied:
`akka-stream` is now defined at the top level

kaiwalyajoshi commented on core/src/main/scala/com/mesosphere/usi/models/package.scala:
```diff
@@ -0,0 +1,45 @@
+
+  sealed trait ReservationState
+
+  object ReservationState {
```
Does it make sense to introduce another state such as `Stale`, `Obsolete` or `Expired` for cases where `TASK_GONE_BY_OPERATOR` can invalidate a previously valid reservation on a lost agent for example?

timcharper replied:
Yeah, totally, let's go in to detail on this later on.

kaiwalyajoshi commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -1,13 +1,65 @@
+ * 1) SpecEvents - Used to replicate the specification state from the framework to the USI scheduler
+ * 2) MesosEvents - Events from Mesos; offers, task status updates, etc.
+ *
+ * Output:
```
If I'm not mistaken, `USIStateEvents` and `MesosCalls` collectively make up a `Frame` here. If so it would be nice to depict that in the diagram and in the writeup here.

timcharper replied:
They make up a FrameResult, or, all of the effects since processing some incoming event (spec updated event or mesos event)

timcharper commented:
@alenkacz  I would recommend you just focus on the overall interface. The details aren't important right now. Their place holders and places where logic will go in the future.

Commit:
More renames, clarifications

- FrameResult is now FrameEffects
- Separate out files
- Update docs. Again.

Commit:
Use test-utils AkkaUnitTest

Commit:
Apply results -> effects naming in SchedulerLogicGraph

Remove extra numbers source
other renames

zen-dog commented on core-models/src/main/scala/com/mesosphere/usi/core/models/RunSpec.scala:
```diff
@@ -0,0 +1,7 @@
+  * Specification used to launch a [[PodSpec]]
+  * @param cpus number of cpus requested by the pod
+  */
+case class RunSpec(cpus: Int)
```
```suggestion
case class RunSpec(cpus: Double)
```

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -1,13 +1,65 @@
+ * 2) MesosEvents - Events from Mesos; offers, task status updates, etc.
+ *
+ * Output:
+ * 1) USIStateEvents - Used to replicate the state of pods, agents, and reservations to the framework
```
Nit: Throughout this review, I kept reading `USIStateEvent` as `USIS-tate-events` where `USIS` sounds somewhat like `ISIS` 😱  I know USI is an abbreviation but could we rename it to `UsiStateEvent` or even just `SchedulerStateEVent`? Prefixes usually have value but in this case, we could also omit it altogether.

jeschkies replied:
Yeah, please not too many caps. I once worked on an `EMLVWProcessorImplFactory`. Why not simply `StateEvents`?

timcharper replied:
Remnants of comments from before renames :(

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -1,13 +1,65 @@
+ * 1) USIStateEvents - Used to replicate the state of pods, agents, and reservations to the framework
+ * 2) MesosCalls - Actions, such as revive, kill, accept offer, etc., used to realize the specification.
+ *
+ * Fully wired, the graph looks like this at a high-level view:
```
Nice graphics 👏

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+  * effects.
+  */
+private[core] class SchedulerLogic {
+  private var frame: Frame = Frame(Map.empty, Map.empty, Map.empty)
```
Wdyt about moving this 3x `Map.empty, Map.empty, Map.empty` to the Frame's constructor as default values?

jeschkies replied:
Or `Frame.empty` similar to `FrameEffects.empty`?

timcharper replied:
sure

timcharper replied:
[ ] - DOING

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+    (goal == Goal.Running) && podRecord.isEmpty
+  }
+
+  private[core] def computeNextStateForPods(frame: Frame)(changedPodIds: Set[PodId]): FrameEffects = {
```
Please add documentation (with the focus on state transitions)

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+      frame.podSpecs.get(podId) match {
+        case None =>
+          // TODO - this should be spurious if the podStatus is non-terminal
+          if (frame.podStatuses.get(podId).exists { status =>
```
I'm not sure I understand what case is handled here - changed `podId` isn't part of the `podSpecs` but has a `podStatus` so we handle it as a terminal or unreachable?

timcharper replied:
It's correct. We clean up terminal or unreachable statuses if the podSpec is removed, and the status is terminal or unreachable.

zen-dog commented on core/src/main/scala/com/mesosphere/usi/models/SpecEvent.scala:
```diff
@@ -0,0 +1,21 @@
+  */
+sealed trait SpecUpdated extends SpecEvent
+
+case class PodSpecUpdated(id: PodId, newState: Option[PodSpec]) extends SpecUpdated
```
So if the `newState == None` we infer that the podSpec was deleted? Wdyt about making it a `PodSpecDeleted(id: PodId)`?

timcharper replied:
I prefer the Option

zen-dog replied:
But, but... `PodSpecDeleted` is so much clearer and easier to read.

zen-dog replied:
also, think about logging. grepping for all deleted podSpecs is easier with `PodSpecDeleted(...)` then it is with `PodSpecUpdated(???, None)`

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+  }
+
+  /**
+    * Process the modification of some podSpec.
```
I believe this comment belongs to the `computeNextStateForPods` method below

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+    * @return The effects of launching this offer
+    */
+  private def matchOffer(offer: Mesos.Event.Offer, pendingLaunchPodSpecs: Seq[PodSpec]): FrameEffects = {
+    var effects = FrameEffects.empty
```
Wdyt about moving this already into a dedicated `OfferMatcher` class with some interface and one (naive) first-come-first-served implementation? This would show where we want to go with it eventually.

timcharper replied:
yeah, not opposed to doing that later.

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/Frame.scala:
```diff
@@ -0,0 +1,61 @@
+    effects.stateEvents.foreach {
+      case recordChange: PodRecordUpdated =>
+        recordChange.newRecord match {
+          case Some(newRecord) =>
```
It's much more readable to have `PodRecordUpdated` and `PodRecordDelete` than inferring it from the `Option`

timcharper replied:
disagree

zen-dog replied:
![](https://media.giphy.com/media/ALalVMOVR8Qw/giphy.gif)

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/Frame.scala:
```diff
@@ -0,0 +1,61 @@
+          case None =>
+            newPodRecords -= recordChange.id
+        }
+      case statusChange: PodStatusUpdated =>
```
Same here

jeschkies commented on core-models/src/main/scala/com/mesosphere/usi/core/models/PodSpec.scala:
```diff
@@ -1,7 +1,15 @@
-  * @param id Id of the specification.
+  * Specification of some Pod that should be launched.
+  *
+  * Pods are launched at-most-once. It is illegal to transition of podSpec from goal terminal to goal running.
```
```suggestion
  * Pods are launched at-most-once. It is illegal to transition of pod spec from goal terminal to goal running.
```

We should agree on one way to spell it. Camel case is fine with me.

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/Frame.scala:
```diff
@@ -0,0 +1,61 @@
+          pod.id -> pod
+        }(collection.breakOut)
+
+        val changedPodIds = podSpecs.keySet ++ newPodsSpecs.keySet
```
Is this correct? In `changed = existing + new` why are existing podSpecs part of the changed ones? Or ist because existing ones are considered removed later?

timcharper replied:
no longer applicable

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/CachedPendingLaunch.scala:
```diff
@@ -0,0 +1,26 @@
+    * @param podIds
+    */
+  def update(frame: Frame, podIds: Set[PodId]): CachedPendingLaunch = {
+    var newPendingLaunch = pendingLaunch
```
What do you think builder `val builder = pendingLaunch.genericBuilder[PodId]` instead? Maybe it's to early to optimize it.

timcharper replied:
I prefer the immutable builder

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/Frame.scala:
```diff
@@ -0,0 +1,61 @@
+import com.mesosphere.usi.core.models.{PodId, PodRecord, PodSpec}
+import com.mesosphere.usi.models._
+
+case class Frame(podSpecs: Map[PodId, PodSpec], podRecords: Map[PodId, PodRecord], podStatuses: Map[PodId, PodStatus]) {
```
Needs some scaladocs 😉

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/FrameEffects.scala:
```diff
@@ -0,0 +1,31 @@
+
+  def withPodStatus(id: PodId, newStatus: Option[PodStatus]): FrameEffects =
+    withStateUpdated(PodStatusUpdated(id, newStatus))
+  def withPodRecord(id: PodId, newRecord: Option[PodRecord]): FrameEffects =
```
This needs some docs. Eg when `newRecord` is empty it means a deletion, right? Should we have a `withoutPodRecord(id)` as well?

timcharper replied:
`withPodRecord` None handles it for now; I'll add docs

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/Scheduler.scala:
```diff
@@ -1,13 +1,65 @@
+ *      \                              |                      |                                 /
+ *       \-----------------------------<        Mesos         <---------------------------------
+ *                                     |                      |
+ *                                     +----------------------+
```
This is really cool. Do you have the PlantUML for it?

timcharper replied:
Thanks! No, PlantUML cannot generate graphs like this. I drew it with Emacs artist-mode

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+  private var frame: Frame = Frame(Map.empty, Map.empty, Map.empty)
+  private var cachedPendingLaunch = CachedPendingLaunch(Set.empty)
+
+  def processSpecEvent(msg: SpecEvent): FrameEffects = {
```
This will be just awesome to test.

zen-dog replied:
Nicely readable logic 👏

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/FrameEffects.scala:
```diff
@@ -0,0 +1,31 @@
+import com.mesosphere.usi.models._
+
+case class FrameEffects(reverseStateEvents: List[USIStateEvent] = Nil, reverseMesosCalls: List[Mesos.Call] = Nil) {
+  lazy val stateEvents = reverseStateEvents.reverse
```
Maybe just take a `Queue` instead of `List`s so there is no need to reverse?

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+    *
+    * @return The effects of launching this offer
+    */
+  private def matchOffer(offer: Mesos.Event.Offer, pendingLaunchPodSpecs: Seq[PodSpec]): FrameEffects = {
```
I'm surprised this method is not receiving a frame. It feels the `pendingLaunchPodsSpecs` should be part of the frame.

timcharper replied:
`pendingLaunchPodsSpecs` is derived state from the frame, and is incrementally recomputed (a cache).

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+    pendingLaunchPodSpecs.foreach { pod =>
+      effects = effects.withPodRecord(pod.id, Some(PodRecord(pod.id, Instant.now(), offer.agentId)))
+    }
+    effects = effects.withMesosCall(Mesos.Call.Accept(offer.offerId, operations = operations))
```
Hm, I know you've said we should concentrate on the interfaces but maybe this is part of it.
```scala
pendingLaunchPodSpecs.foreach { pod =>
  effects += UpdatedPodRecord(...)
}
effetcs += MesosCall(...)
```

Was easier to comprehend. It took me two looks to understand why `effects` was not simply overriden. This really feels like a source of possible bugs. Even a mutable `Effects` object feels nicer such as an effects builder

timcharper replied:
I'll change it to a fold

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+    * @param podIds podIds changed during the last frame
+    * @return
+    */
+  def pruneTaskStatuses(frame: Frame)(podIds: Set[PodId]): FrameEffects = {
```
Hm, it really feels we should return a `Frame` instead.

timcharper replied:
Preference for FrameEffects. See FrameWithEffects abstraction.

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/CachedPendingLaunch.scala:
```diff
@@ -0,0 +1,26 @@
+package com.mesosphere.usi.core
+import com.mesosphere.usi.core.models.PodId
+
+case class CachedPendingLaunch(pendingLaunch: Set[PodId] = Set.empty) {
```
This looks like premature optimization to me. Is there a benchmark showing that filtering a list of pod specs actually takes long? How many pod specs do you need to see the difference and how big is it? Is it on the order of μs? ms?

timcharper replied:
Discussed; well aware of premature optimization arguments. It is obvious that we will need caching mechanisms, and I wanted to work out how they will look early in the development of the SchedulerLogic. Recomputing pending launch states for every message is nonsense. I'd like to keep it.

jeschkies commented on core/src/main/scala/com/mesosphere/usi/models/USIStateEvent.scala:
```diff
@@ -0,0 +1,24 @@
+  * events. They are emitted so that a framework can replicate the state and react to changes in a way that best suits
+  * the framework's domain.
+  */
+sealed trait USIStateEvent
```
```suggestion
sealed trait StateEvent
```
I don't think we would have something non USI related in `com.mesosphere.usi.models`.

timcharper replied:
✅

jeschkies commented on core/src/main/scala/com/mesosphere/usi/models/package.scala:
```diff
@@ -0,0 +1,45 @@
+
+import com.mesosphere.usi.core.models._
+
+package models {
```
Any reason this is not part of `core-models`?

timcharper replied:
Can we move it after?

zen-dog replied:
I moved it. Don't forget to pull

jeschkies reviewed (commented):
I really like it. Some things are not quite clear to me but the `Frame` feels like a good approach.

Any reason some models are not part of `core-models`? It feels like that;s a reason to reject.

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+  }
+
+  /**
+    * Given a frame effects, apply the usi state change effects of the frame effects. Then, update the cache and perform
```
I had to read the first sentence three times and I still don't quite understand it 😊 Maybe a diagram would help?

timcharper replied:
It should be simpler now with FrameWithEffects

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+  */
+private[core] class SchedulerLogic {
+  private var frame: Frame = Frame(Map.empty, Map.empty, Map.empty)
+  private var cachedPendingLaunch = CachedPendingLaunch(Set.empty)
```
Why do we need the pending launches cached here like this? Why is not a `lazy val pendingLaunch...` within the above frame since frames are immutable?

timcharper replied:
Discussed in meeting.

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+    this.cachedPendingLaunch = this.cachedPendingLaunch.update(frame, dirtyPodIds)
+
+    val pruneEffects = SchedulerLogic.pruneTaskStatuses(frame)(dirtyPodIds)
+    frame = frame.applyStateEffects(pruneEffects)
```
This "feels dirty", it's the t**hird time** you update the `private var frame` (starting in the `processSpecEvent ` method call chain) and only by the third time, it's a valid frame (meaning all the effects are there). I would expect to see only one `frame = frame.apply...` call that replaces the current frame with a new one. You can have intermediate frames but I would not update the current one with them. This should make the whole logic more "atomic" and the  `cachedPendingLaunch` could become a `lazy val` within the frame.

zen-dog replied:
...and an even better solution is described in my [comment below](https://github.com/mesosphere/usi/pull/28/files#r253903419)

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,201 @@
+    * @return The events describing state changes as Mesos call intents
+    */
+  def processMesosEvent(event: Mesos.Event): FrameEffects = {
+    val result = SchedulerLogic.handleMesosEvent(frame, cachedPendingLaunch.pendingLaunch)(event)
```
This split into `case class SchedulerLogic` and `object SchedulerLogic` is only necessary (to ease the testing) because `SchedulerLogic` holds the `frame`. If the frame would live outside then both could be united into a set of pure methods: _pass current frame and event -> return new frame and effects_. 
I believe @jeschkies suggested something similar and I support that.

timcharper replied:
Added FrameWithEffects abstraction which handles this more elegantly. Preference to have business logic only return effects if possible.

zen-dog reviewed (dismissed):
I left a few comments at places which IMHO require some reevaluation (especially around `SchedulerLogic`) so let's talk about it. Overall, this is a great PR! I love the BidiFlow and the diagrams (keep them coming please).

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogicGraph.scala:
```diff
@@ -0,0 +1,109 @@
+
+  // This is where the actual (possibly stateful) logic will live
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic = {
+    val schedulerLogic = new SchedulerLogic
```
This is where I would put current `Frame` and make `SchedulerLogic` pure

zen-dog replied:
... or not ;)

timcharper replied:
Added FrameWithEffects; we can shuffle things around further, later, if we need.

Commit:
Introduce FrameWithEffects type and clean up mutable assignments

FrameWithEffects abstraction allows us to method-chain the pipeline for
processing a frame; it handles the accumulation and application of
FrameEffects. It also tracks dirty state throughout the lifecycle of the Frame
evaluation.

Commit:
Put MockMesos in core for now and add Scheduler.connect method

Commit:
Get rid of USI in the names

Commit:
Replace vars with folds and other functional things

Commit:
Docs

timcharper commented:
@jeschkies Not sure if I agree this should be rejected because models are in wrong path. This is a mock implementation and we still preserve the ability to change things in the future. If you feel strongly please add a commit and move them.

Commit:
docs etc

jeschkies committed:
Update core-models/src/main/scala/com/mesosphere/usi/core/models/Goal.scala

Co-Authored-By: timcharper <tharper@mesosphere.io>

Commit:
Update core-models/src/main/scala/com/mesosphere/usi/core/models/RunSpec.scala

Co-Authored-By: timcharper <tharper@mesosphere.io>

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/FakeMesos.scala:
```diff
@@ -0,0 +1,23 @@
+import com.mesosphere.usi.core.models.AgentId
+import com.mesosphere.usi.models.Mesos
+
+object FakeMesos {
```
Should we move this to `test-utils`?

zen-dog replied:
I don't think so.

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,237 @@
+    val newDirty = dirtyPodIds ++ newEffects.reverseStateEvents.iterator.collect {
+      case podEvent: PodStateEvent => podEvent.id
+    }
+    copy(frame = frame.applyStateEffects(effects), dirtyPodIds = newDirty, effects = effects ++ newEffects)
```
Shouldn't the original effects be dropped since they are already applied?

zen-dog replied:
Good question. However, of these effects, only `stateEvents` (and not e.g. `mesosCalls`) are applied. So I guess we should keep the later and remove the former?

timcharper replied:
We have to aggregate the events so we can publish them downstream

jeschkies reviewed (commented):
I like the new `FrameWithEffects` but am unsure about `FrameWithEffects.applyEffects`. Maybe I misunderstand something.

zen-dog committed:
Moved all models to `core-models` and fixed the imports

zen-dog commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,235 @@
+    val newDirty = dirtyPodIds ++ newEffects.reverseStateEvents.iterator.collect {
+      case podEvent: PodStateEvent => podEvent.id
+    }
+    copy(frame = frame.applyStateEffects(effects), dirtyPodIds = newDirty, effects = effects ++ newEffects)
```
```suggestion
    copy(frame = frame.applyStateEffects(newEffects), dirtyPodIds = newDirty, effects = effects ++ newEffects)
```
This above looks like a bug to me.

zen-dog replied:
Also, I believe we conflate two things in _Effects_:
1) things that change the Frame
2) side effects (like Mesos calls)
Could we call the 2) - `Actions` to make their purpose clearer?

timcharper replied:
I could go with intents, perhaps? Let's talk

Commit:
Update core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala

Co-Authored-By: timcharper <tharper@mesosphere.io>

jeschkies commented on core/src/main/scala/com/mesosphere/usi/core/SchedulerLogic.scala:
```diff
@@ -0,0 +1,235 @@
+        .process { f =>
+          SchedulerLogic.computeNextStateForPods(f.frame)(f.dirtyPodIds)
+        }
+    }
```
Here's my suggestion to flatten it

```scala
val (frameWithSpecEventsApplied, dirityPods) = applySpecEvent(currentFrame, msg)
val FrameIntents(stateUpdates, calls) = SchedulerLogic.computeNextStateForPods(frameWithSpecEventsApplied)(dirtyPodIds)

val finalFrame = applyIntents(stateUpdates)
currentFrame = finalFrame
FrameIntents(stateUpdates, calls)
```
The names are a little verbose.

timcharper replied:
I really prefer the method chaining approach; it gets rid of all of these assignments and precludes us from accidentally returning the wrong version of the frame; further, the chained abstraction separates out the concern of tracking dirty podIds, which I really like. Can you help me understand why you think this is better?

Commit:
Rename Frame to SchedulerLogicState (and other associated renames)

The concept of frame was too muddy. Frame is specifically being defined as the
state of the schedulerLogic between each handling of an incoming event. The
frame state evolves as we handle various pieces of logic in response to that
incoming event.

We also double-down on separating, even though it is just semantically, the
difference between intents and events. Intents are things not yet applied.
Events are things that were applied and we're notifying you about. StateEvents
can be returned as an intent to be applied (since per the event-sourcing
pattern we evolve our state via the same events that we emit and persist), but
after they are applied, they're referred to as events and not intents.

Commit:
Further push frame -> state renames

Commit:
Note on events and intents in the context of SchedulerLogic

Commit:
note about frames; add headings

Commit:
Separate out SpecificationState from SchedulerLogicState

Commit:
Run over and refresh model docs
